packages/protocol/graphics.ts

```diff
@@ -652,63 +663,48 @@ export async function getFirstMeaningfulPaint(limit: number = 10) {
   }
 }
 
-// precache this many milliseconds
-const precacheTime = 5000;
-// startTime of the currently running precacheScreenshots() call
-let precacheStartTime = -1;
+////////////////////////////////////////////////////////////////////////////////////////////////
+// The callbacks below allow external code to observe specific events.
+// This design pattern is a short term strategy to remove external imports from this package.
+//
+// TODO We should revisit this as part of a larger architectural redesign (#6932).
+////////////////////////////////////////////////////////////////////////////////////////////////
 
-export function precacheScreenshots(startTime: number): UIThunkAction {
-  return async (dispatch, getState) => {
-    const recordingDuration = getRecordingDuration(getState());
-    if (!recordingDuration) {
-      return;
-    }
-
-    startTime = snapTimeForPlayback(startTime);
-    if (startTime === precacheStartTime) {
-      return;
-    }
-    if (startTime < precacheStartTime) {
-      dispatch(setPlaybackPrecachedTime(startTime));
-    }
-    precacheStartTime = startTime;
-
-    const endTime = Math.min(startTime + precacheTime, recordingDuration);
-    for (let time = startTime; time < endTime; time += snapInterval) {
-      const index = mostRecentIndex(gPaintPoints, time);
-      if (index === undefined) {
-        return;
-      }
+export interface Canvas {
+  gDevicePixelRatio: number;
+  height: number;
+  left: number;
+  scale: number;
+  top: number;
+  width: number;
+}
 
-      const paintHash = gPaintPoints[index].paintHash;
-      if (!screenshotCache.hasScreenshot(paintHash)) {
-        const graphicsPromise = getGraphicsAtTime(time, true);
+let onMouseDownEvents: (events: MouseEvent[]) => void;
+export function setMouseDownEventsCallback(callback: typeof onMouseDownEvents): void {
```

Comment:
agreed that this is temporary, but it might feel a bit better to expose a simpler event listener interface through `addListener`, `emitEvent`.

```js
function addListener(event, callback) {
  gCallbacks[event]?.push(callback);
}
function emitEvent(event, ...args) {
  for (const callback in gCallbacks[event]) {
    callback(...args)
  }
}
```

Reply:
FWIW we have at least two existing `EventEmitter` classes in this codebase.  One is in `devtools/`, plain JS, and more extensive behavior.  The other is in `ui/utils/`, is TS, and same basic API but simpler internals.

(but neither of them gives fully typed events, which is why I was debating pulling in a _third_ implementation at some point...)

Reply:
That's a modified "Hungarian Notation" syntax (a concept that goes _allll_ the way back to old-school Microsoft C/C++ days):

- `g` for global variables
- `m` for class members

etc

Reply:
Fair enough but...why are we using it? These concepts aren't any more "global" than any of our other module level exports :D

Reply:
TBH I thought the "g" had something to do with "graphics" bc' it was in that file 😆

Reply:
You'll see that notation used a lot more in the backend. I suspect the "how did it get that way" has to do with who originally wrote it and whether or not they preferred that style, but does not have a lot of firm reasons beyond that.

Reply:
Thanks for the background info! 😄

Reply:
> agreed that this is temporary, but it might feel a bit better to expose a simpler event listener interface through `addListener`, `emitEvent`.

I'm not opposed to this, but I've gotten bogged down in TypeScript generics trying to write a strongly typed event dispatcher API and I'd rather punt on that as far as this PR goes.

Reply:
Yeah. Early in my career I used it because I was writing C++/Java and that's how the example projects I learned from wrote things.  Later on I dropped it because it didn't seem as useful (and especially as I got into JS and more FP-inspired styles).